spec/save_operation_callbacks_spec.cr

```diff
@@ -74,14 +83,49 @@ private class SaveLineItemSub < SaveLineItemBase
   end
 end
 
+private class SaveOperationWithCallbacks < Post::SaveOperation
+  include TestableOperation
+
+  before_save :set_title
+  before_save { mark_callback("before_save_in_a_block") }
+
+  after_save :notify_save_complete
+  after_save do |saved_post|
+    mark_callback("after_save_in_a_block with #{saved_post.title}")
+  end
+
+  after_commit :notify_commit_complete
+  after_commit do |saved_post|
+    mark_callback("after_commit_in_a_block with #{saved_post.title}")
+  end
+
+  private def set_title
+    mark_callback("before_save_update_title")
+    title.value = "Saved Post"
+  end
+
+  private def notify_save_complete(saved_post)
+    mark_callback("after_save_notify_save_complete with #{saved_post.title}")
+  end
+
+  private def notify_commit_complete(saved_post)
+    mark_callback("after_commit_notify_commit_complete with #{saved_post.title}")
+  end
+end
+
+private class UpdateOperationWithSkipCallbacks < SaveOperationWithCallbacks
+  permit_columns :title
+  skip_before_save :set_title
+  skip_after_save :notify_save_complete
+  skip_after_commit :notify_commit_complete
+end
+
 describe "Avram::SaveOperation callbacks" do
-  it "does not run after_* callbacks if just validating" do
+  it "does not run any callbacks if just validating" do
```

Comment:
This is because `save` must be called now to kick off the `before_save` callbacks.